### PR TITLE
Managed streaming Ingest 

### PR DESCRIPTION
**Pull request by AsafMah**
* Adds ManagedStreamingIngest - works similarly to other sdks:
    * Ingesting from blob forwards to queued
    * Ingesting >4mb forwards to queued
    * Otherwise, streaming is tried
    * Multiple transient failures dealt with exponential backoff, and after 3 falls back to queued
    * Well tested and covered, both e2e and unit
* Removes RequiredMapping as it's no longer a requirement from the engine
* Fixes several warnings

### CODE REVIEW

**Comment by AsafMah** [comment]
@element-of-surprise I don't understand, that's what I'm doing right now in the code - poll the results and wait 100ms between polls,

Not sure how i can do it otherwise

File: kusto/ingest/file_options.go [hunk 1]
```diff
@@ -91,14 +93,65 @@ func (o option) Run(p *properties.All, clientType ClientScope, sourceType Source
...
+// Compress sets whether to compress the data.
+func Compress(compress bool) FileOption {
+	return option{
+		run: func(p *properties.All) error {
+			p.Source.DontCompress = !compress
+			return nil
+		},
+		clientScopes: QueuedClient | StreamingClient | ManagedClient,
+		sourceScope:  FromFile | FromReader | FromBlob,
+		name:         "FlushImmediately",
```
Comment (AsafMah) [inline kusto/ingest/file_options.go:131]: Should be Compress + don't from blob

Comment (AsafMah) [reply kusto/ingest/file_options.go:131]: replace with DontCompress

File: kusto/ingest/file_options.go [hunk 2]
```diff
@@ -222,6 +275,7 @@ func DeleteSource() FileOption {
 			p.Source.DeleteLocalSource = true
 			return nil
 		},
+		// This option is not available to ManagedStreaming, due to unexpected behavior when using streaming vs queued ingestion.
 		clientScopes: QueuedClient | StreamingClient,
```
Comment (AsafMah) [inline kusto/ingest/file_options.go:279]: Figure out if it works for managed

File: kusto/ingest/ingest.go [hunk 3]
```diff
@@ -158,13 +165,18 @@ func (i *Ingestion) FromFile(ctx context.Context, fPath string, options ...FileO
 // ingested after all data in the reader is processed. Content should not use compression as the content will be
 // compressed with gzip. This method is thread-safe.
 func (i *Ingestion) FromReader(ctx context.Context, reader io.Reader, options ...FileOption) (*Result, error) {
-	result, props, err := i.prepForIngestion(ctx, options, FromReader)
+	return i.fromReaderProps(ctx, reader, options, i.newProp())
+}
+
+// fromReaderProps is an internal function to allow managed streaming to pass a properties object to the ingestion.
+func (i *Ingestion) fromReaderProps(ctx context.Context, reader io.Reader, options []FileOption, props properties.All) (*Result, error) {
```
Comment (AsafMah) [inline kusto/ingest/ingest.go:172]: rename to fromReader

File: kusto/ingest/managed.go [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+	return &Managed{
+		queued:    queued,
+		streaming: streaming,
+	}, nil
+}
+
+func (m *Managed) FromFile(ctx context.Context, fPath string, options ...FileOption) (*Result, error) {
+	props := m.newProp()
+	file, err := prepFile(fPath, &props, options, ManagedClient)
```
Comment (AsafMah) [inline kusto/ingest/managed.go:50]: check the file here too

**Review by AsafMah — COMMENTED** [review]
bad job!

File: kusto/ingest/managed.go (hunk 4 above)
Comment (AsafMah) [reply kusto/ingest/managed.go:49]: Done (in prepfile)

File: kusto/ingest/ingest.go (hunk 3 above)
Comment (AsafMah) [reply kusto/ingest/ingest.go]: done

File: kusto/ingest/file_options.go (hunk 2 above)
Comment (AsafMah) [reply kusto/ingest/file_options.go]: Done

File: kusto/ingest/file_options.go (hunk 1 above)
Comment (AsafMah) [reply kusto/ingest/file_options.go]: Done

File: kusto/ingest/internal/resources/resources.go [hunk 5]
```diff
@@ -167,6 +171,8 @@ func (m *Manager) renewResources() {
 // AuthContext returns a string representing the authorization context. This auth token is a temporary token
 // that can be used to write a message via ingestion.  This is different than the ADAL token.
 func (m *Manager) AuthContext(ctx context.Context) (string, error) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
 	if m.kustoTokenCacheExpiration.After(time.Now().UTC()) {
 		return m.kustoToken.AuthContext, nil
 	}
```
[the remark this answers is not in the document]
Comment (AsafMah) [reply kusto/ingest/internal/resources/resources.go:180]: Done

File: kusto/ingest/file_options.go [hunk 6]
```diff
@@ -91,14 +93,65 @@ func (o option) Run(p *properties.All, clientType ClientScope, sourceType Source
...
+			p.Source.DontCompress = !compress
+			return nil
+		},
+		clientScopes: QueuedClient | StreamingClient | ManagedClient,
+		sourceScope:  FromFile | FromReader | FromBlob,
+		name:         "FlushImmediately",
+	}
+}
+
+func BackOff(off *backoff.ExponentialBackOff) FileOption {
```
[the remark this answers is not in the document]
Comment (AsafMah) [reply kusto/ingest/file_options.go]: Done

**Review by alonadam — APPROVED** [review]

**Review by yogilad — APPROVED** [review]

**Pull request merged**